grammar.js

```diff
@@ -24,497 +39,1236 @@ module.exports = grammar({
-      $.identifier,
-      ':',
-      $._pattern
+    _type_equation: $ => seq(
+      choice('=', ':='),
+      optional('private'),
+      $._type_expression
     ),
 
-    shorthand_labeled_parameter: $ => seq(
-      choice('~', '?'),
-      $.identifier
+    variant_declaration: $ => seq(
+      optional('|'),
+      sep1('|', $.constructor_declaration)
     ),
 
-    type_constructor_definition: $ => seq(
-      optional($.type_parameters),
-      $._type_identifier,
-      optional($.type_equation),
-      optional(seq(
-        '=',
-        $.record_declaration
+    constructor_declaration: $ => seq(
+      choice(
+        $.constructor_name,
+        seq('[', ']'),
+        parenthesize('::')
+      ),
+      optional(choice(
+        seq(
+          'of',
+          choice(
+            sep1('*', $._simple_type_expression),
+            $.record_declaration
+          )
+        ),
+        seq(':', $._type_expression),
+        seq('=', $.constructor_path)
       )),
-      repeat($.type_constraint)
-    ),
-
-    type_parameters: $ => repeat1(seq(
-      optional(choice('+', '-')),
-      "'",
-      $._type_identifier
-    )),
-
-    type_equation: $ => seq(
-      '=',
-      $._type
+      repeat($.attribute)
     ),
 
     record_declaration: $ => seq(
       '{',
-      sep(';', $.field_declaration),
-      optional(';'),
+      sep1(seq(';', repeat($.attribute)), $.field_declaration),
+      optional(seq(';', repeat($.attribute))),
       '}'
     ),
 
     field_declaration: $ => seq(
       optional('mutable'),
-      $.identifier,
+      $.field_name,
       ':',
-      $._type
+      $._poly_type_expression,
     ),
 
     type_constraint: $ => seq(
       'constraint',
-      $._type_identifier,
+      $.type_variable,
       '=',
-      $._type
+      $._type_expression
     ),
 
-    // Module types
+    exception_definition: $ => seq(
+      'exception',
+      $.constructor_declaration,
+      repeat($.item_attribute)
+    ),
 
-    _module_type: $ => choice(
-      $.signature,
-      $._type_identifier,
-      $.functor_module_type,
-      $.shorthand_functor_module_type,
-      $.qualified_type_identifier
+    module_definition: $ => seq(
+      'module', optional($._extension_attribute), optional('rec'),
+      sep1(seq('and', repeat($.attribute)), $.module_binding)
+    ),
+
+    module_binding: $ => seq(
+      $.module_name,
+      repeat($.module_parameter),
+      optional(seq(':', $._module_type)),
+      optional(seq('=', $._module_expression)),
+      repeat($.item_attribute)
+    ),
+
+    module_parameter: $ => parenthesize(optional(seq(
+      choice($.module_name, alias('_', $.module_name)),
+      ':',
+      $._module_type
+    ))),
+
+    module_type_definition: $ => seq(
+      'module', 'type',
+      optional($._extension_attribute),
+      $.module_type_name,
+      optional(seq('=', $._module_type)),
+      repeat($.item_attribute)
+    ),
+
+    open_statement: $ => seq(
+      /open!?/,
```

Comment:
I wanted this to be `choice('open', 'open!')`, but that resulted in parsing errors I couldn't solve.